### ISSUE TITLE
bug: `encodeEventTopics` doesn't encode args correctly for some events

### ISSUE THREAD

**Issue by kdembler** [wagmi-dev/viem#485]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

latest

### Current Behavior

`encodeEventTopics` that's used among others by `getLogs` doesn't properly encode arguments into topics for some events. Take a look at this code snippet:

https://github.com/wagmi-dev/viem/blob/1c539f82dd49d811ee771a030f2597eb1da86225/src/utils/abi/encodeEventTopics.ts#L62-L79

Let's assume `args` is an object, not an array. Firstly, `args_` is created by mapping over all `abiItem.inputs` - **this will include both indexed and not indexed args.** If we imagine an event that has 6 arguments, and only last of them is indexed and included in `args`, this can produce an example `const args_ = [undefined, undefined, undefined, undefined, undefined, 'value']`.

Next, `topics` is populated by mapping `abiItem.inputs`, **but this time only `indexed` ones.**. Then array index is used to access `args_`, but because `args_` is created from all the inputs, there are index mismatches. And instead using `'value'` as argument, `undefined` is accessed.

In conclusion, this function will work only if all the indexed event arguments are placed before any non-indexed arguments in event signature, which doesn't have to be the case. Hopefully my explanation makes sense, let me know if there's any confusion

**Comment by kdembler** [comment]
Opened a fix: #487

**Comment by jxom** [comment]
Good catch! Will take a look tomorrow. 🙏
